src/Uno.UI/UI/Xaml/ResourceDictionary.cs

```diff
@@ -3,98 +3,91 @@
 using System.Threading;
 using Uno.UI;
 using Uno.Extensions;
+using System.ComponentModel;
 
 namespace Windows.UI.Xaml
 {
-	public partial class ResourceDictionary : DependencyObject
+	public partial class ResourceDictionary : DependencyObject, IDependencyObjectParse, IDictionary<object, object>
 	{
 		private readonly Dictionary<object, object> _values = new Dictionary<object, object>();
-		private int _resolvingDepth = 0;
 
 		public ResourceDictionary()
 		{
 		}
 
-		public static Func<string, object> DefaultResolver { get; set; }
-
+		private Uri _source;
 		public Uri Source
 		{
-			get;
-			set;
+			get => _source;
+			set
+			{
+				if (!IsParsing) // If we're parsing, the Source is being set as a 'FYI', don't try to resolve it
+				{
+					var sourceDictionary = ResourceResolver.RetrieveDictionaryForSource(value);
+
+					CopyFrom(sourceDictionary);
+				}
+
+				_source = value;
+			}
 		}
 
 		public IList<ResourceDictionary> MergedDictionaries { get; } = new List<ResourceDictionary>();
 
-		public IDictionary<object, object> ThemeDictionaries { get; } = new Dictionary<object, object>();
-
+		private IDictionary<object, object> _themeDictionaries;
+		public IDictionary<object, object> ThemeDictionaries { get => _themeDictionaries = _themeDictionaries ?? new ResourceDictionary(); }
 		public object Lookup(object key)
 		{
 			object value;
 
-			var keyName = key.ToString();
+			var keyName = key;
 			if (!_values.TryGetValue(keyName, out value))
 			{
-				return DefaultResolver?.Invoke(keyName);
+				return null;
 			}
 
 			return value;
 		}
 
 		public bool HasKey(object key)
 		{
-			var keyName = key.ToString();
+			var keyName = key;
 
-			return _values.ContainsKey(keyName) || DefaultResolver?.Invoke(keyName) != null;
+			return _values.ContainsKey(keyName);
 		}
 
 		public bool Insert(object key, object value)
 		{
-			_values[key.ToString()] = value;
+			_values[key] = value;
 			return true;
 		}
 
-		public bool Remove(object key) => _values.Remove(key.ToString());
+		public bool Remove(object key) => _values.Remove(key);
 
-		public bool Remove(KeyValuePair<object, object> key) => _values.Remove(key.Key.ToString());
+		public bool Remove(KeyValuePair<object, object> key) => _values.Remove(key.Key);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Rename parameter 'key' to 'item' to match the interface declaration.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=4826539)